lib/resources/virtualization.rb

```diff
@@ -0,0 +1,246 @@
+# encoding: utf-8
+# author: Takaaki Furukawa
+
+require 'hashie/mash'
+
+module Inspec::Resources
+  class Virtualization < Inspec.resource(1) # rubocop:disable Metrics/ClassLength
+    name 'virtualization'
+    desc 'Use the virtualization InSpec audit resource to test the virtualization platform on which the system is running'
+    example "
+      describe virtualization.system do
+        it { should eq 'docker' }
+      end
+
+      describe virtualization do
+        its('role') { should eq 'guest' }
+      end
+
+      describe virtualization.systems[:docker] do
+        it { should eq 'guest' }
+      end
+
+      if virtualization.system == 'docker'
+        describe file('/var/tmp/foo') do
+          it { should be_file }
+        end
+      end
+    "
+
+    def initialize
+      unless inspec.os.linux?
+        skip_resource 'The `virtualization` resource is not supported on your OS yet.'
+      end
+    end
+
+    def [](name)
+      # convert string to symbol
+      name = name.to_sym if name.is_a? String
+      data = collect_data_linux
+      data[name]
+    end
+
+    # add helper methods for easy access of properties
+    # allows users to use virtualization.role, virtualization.system, virtualization.systems
+    %w{role system systems}.each do |property|
+      define_method(property.to_sym) do
+        data = collect_data_linux
+        data[property.to_sym]
+      end
+    end
+
+    def params
+      collect_data_linux
+    end
+
+    def to_s
+      'Virtualization Detection'
+    end
+
+    private
+
+    def lxc_version_exists?
+      inspec.command('lxc-version').exist?
+    end
+
+    def docker_exists?
+      inspec.command('docker').exist?
+    end
+
+    def nova_exists?
+      inspec.command('nova').exist?
+    end
+
+    def collect_data_linux # rubocop:disable Metrics/MethodLength, Metrics/PerceivedComplexity, Metrics/CyclomaticComplexity, Metrics/AbcSize
```

Comment:
Wow, one heck of a method!

My main concern is that it tests for EVERY virtualization possibility, even if one has already matched. For example, if we detect the host is a Xen host, we'll still test for vbox, openstack, etc.

I wonder if it would make more sense to break apart each of the detections into smaller methods that return true if it matched, false if it didn't... example:

```
def detect_xen
  return false unless inspec.file('/proc/xen').exist?
  @virtualization_data[:system] = 'xen'
  # etc...
  true
end
```

Then `collect_data_linux` could be written like so:

```
def collect_data_linux
  # cache data in an instance var to avoid doing multiple detections for a single test
  @virtualization_data ||= Hashie::Mash.new
  return @virtualization_data unless @virtualization_data.empty?

  # each detect method will return true if it matched and was successfully
  # able to populate @virtualization_data with stuff.
  return if detect_xen
  return if detect_openstack
  return if detect_foo
end
```

I think this would make the code a bit more readable/manageable, and will also make it easier to test with each system detection being its own method.

What do you think, @tkak?